<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="img/favicon.ico">
  <title>Advanced run ORFold - ORFmine Documentation</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="css/theme.css" />
  <link rel="stylesheet" href="css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  <link href="css/extra.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Advanced run ORFold";
    var mkdocs_page_input_path = "Run_orfold_advanced.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="js/jquery-2.1.1.min.js" defer></script>
  <script src="js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href="." class="icon icon-home"> ORFmine Documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="index.html">Home</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">ORFmine</span></p>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="orfmine_installation.html">Installation</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="orfmine_quickstart.html">Quick start</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="orfmine_citations.html">Citations</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">ORFtrack</span></p>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="#">ORFtrack presentation</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="orftrack_description.html">Aims and general description</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="orftrack_orfdef.html">ORF definition</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="orftrack_annotation.html">ORF annotation</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="orftrack_overlap.html">Overlap definition</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="orftrack_orf_extraction.html">ORF extraction</a>
                </li>
    </ul>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="#">Usage</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="orftrack_input.html">Inputs</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="orftrack_run.html">ORF annotation with ORFtrack</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="orfget_run.html">ORF extraction with ORFget</a>
                </li>
    </ul>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="#">Parameters</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="orftrack_param.html">ORFtrack parameters</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="orfget_param.html">ORFget parameters</a>
                </li>
    </ul>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">ORFold</span></p>
                <ul class="current">
                    <li class="toctree-l1"><a class="reference internal" href="#">ORFold presentation</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="Objective_orfold.html">Objective</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="How_it_works_orfold.html">How it works?</a>
                </li>
    </ul>
                    </li>
                    <li class="toctree-l1 current"><a class="reference internal current" href="#">Usage</a>
    <ul class="current">
                <li class="toctree-l2"><a class="reference internal" href="Run_orfold.html">Basic run ORFold</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="Plot_orfold.html">Plots with ORFplot</a>
                </li>
                <li class="toctree-l2 current"><a class="reference internal current" href="Run_orfold_advanced.html">Advanced run ORFold</a>
    <ul class="current">
    <li class="toctree-l3"><a class="reference internal" href="#advanced-run">Advanced run:</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#mapping-of-the-fold-potential-and-the-disorder-and-aggregation-propensities-along-the-genome-of-an-organism">Mapping of the fold potential and the disorder and aggregation propensities along the genome of an organism</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#dealing-with-multiple-files-at-the-same-time">Dealing with multiple files at the same time:</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#running-orfold-on-subsets-of-randomly-selected-sequences">Running ORFold on subsets of randomly selected sequences</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
    </ul>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="#">Parameters</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="parameters_orfold.html">ORFold parameters</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="parameters_orfplot.html">ORFplot parameters</a>
                </li>
    </ul>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href=".">ORFmine Documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".">Docs</a> &raquo;</li>
    
      
        
          <li>Usage &raquo;</li>
        
      
        
          <li>ORFold &raquo;</li>
        
      
    
    <li>Advanced run ORFold</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="running-orfold">Running ORFold:<a class="headerlink" href="#running-orfold" title="Permanent link">#</a></h1>
<h2 id="advanced-run">Advanced run:<a class="headerlink" href="#advanced-run" title="Permanent link">#</a></h2>
<h3 id="mapping-of-the-fold-potential-and-the-disorder-and-aggregation-propensities-along-the-genome-of-an-organism">Mapping of the fold potential and the disorder and aggregation propensities along the genome of an organism<a class="headerlink" href="#mapping-of-the-fold-potential-and-the-disorder-and-aggregation-propensities-along-the-genome-of-an-organism" title="Permanent link">#</a></h3>
<p>In the previous section we presented how to launch ORFold on a set
of amino acid sequences stored in a FASTA file. However, 
the originality of ORFold relies on the fact that the user can manually 
inspect the distribution of the properties estimated with ORFold (fold potential,
and disorder and aggregation propensities) along a genome of interest. 
In this case, the user must provide the genome annotation file (GFF) along with 
the input FASTA file. ORFold will return new GFF files (one per studied property)
that contain for the ORFs provided in the input FASTA file, their corresponding 
property (fold potential, disorder or aggregation propensities). The 
values are stored in the column #9 of the output GFF files. The GFF files can be subsequently
uploaded on a genome viewer such as IGV [1].</p>
<p>The input GFF file must be given with the <strong>-gff</strong> option as follows:</p>
<pre><code>orfold -fna sequences.fasta -options HIT -gff sequences.gff 
</code></pre>
<p>ORFold generates a <strong>sequences.tab</strong> file containing the fold potential, and the 
disorder and aggregation propensities of each sequence present in the input FASTA file.
Additionally, ORFold produces three new GFF files:</p>
<ol>
<li>sequences_HCA.gff</li>
<li>sequences_IUPRED.gff</li>
<li>sequences_TANGO.gff</li>
</ol>
<p>The output GFF files are identical to the one provided by the user except that for the sequences present
at the same time in the FASTA and the GFF files that were given as inputs, the column #9
is now replaced by the fold potential, or the disorder or aggregation propensities estimated by ORFold.
That way, the corresponding sequences can be colored according to these values on
a genome viewer, thereby enabling the visual inspection of these properties along
the input genome. Notice that on IGV, blue indicates low values (for all mapped properties) 
while red indicates high values. </p>
<p><img alt="HCA Scale" src="img/mapping/Scale.png" /><br>
<em>Figure 1: Color scale for the HCA score values 
</em></p>
<div class="admonition note">
    <p class="first admonition-title">
        Note
    </p>
    <p class="last">

Notice that the ID 
of the sequences given in the FASTA file (i.e. annotation after the ">" in 
the FASTA file) must be strictly identical to those of the corresponding sequences 
in the GFF file (i.e. ID indicated in the column #3 of the GFF file).  

    </p>
</div>

<h3 id="dealing-with-multiple-files-at-the-same-time">Dealing with multiple files at the same time:<a class="headerlink" href="#dealing-with-multiple-files-at-the-same-time" title="Permanent link">#</a></h3>
<p>ORFold can handle multiple input files (FASTA and GFF) and will associate 
the FASTA and GFF files according to the following rules:</p>
<ol>
<li>
<p>If the user provides the same number of FASTA and GFF files, ORFold associates them 
   based on their root name, no matter the order of the files.</p>
<pre><code>orfold -fna sequences_Y.fasta sequences_X.fasta -options H -gff sequences_X.gff sequences_Y.gff
</code></pre>
<p>In this case, ORFold associates:</p>
<ul>
<li>sequence_Y.fasta with sequence_Y.gff</li>
<li>sequence_X.fasta with sequence_X.gff </li>
</ul>
<p>&nbsp;</p>
</li>
<li>
<p>I If the user provides the same number of FASTA and GFF files,
   but their root names are not identical, ORFold associates them 
   according to the order of the files in the command line.</p>
<pre><code>orfold -fna sequences_Y.fasta sequences_X.fasta -options H -gff sequences_A.gff sequences_B.gff
</code></pre>
<p>In this case, ORFold associates:</p>
<ul>
<li>sequence_Y.fasta with sequence_A.gff</li>
<li>sequence_X.fasta with sequence_B.gff</li>
</ul>
<p>&nbsp;</p>
</li>
<li>
<p>If the user provides multiple FASTA files and only one GFF file, then:</p>
<ul>
<li>
<p>If the name of the GFF file matches with the name of one of the FASTA files,
  the two files are associated, while the other FASTA files are not associated
  to the input GFF file.</p>
<pre><code>orfold -fna sequences_Y.fasta sequences_X.fasta -options H -gff sequences_X.gff
</code></pre>
<p>ORFold associates:</p>
<ul>
<li>sequences_X.fasta with sequences_X.gff</li>
<li>sequences_Y.fasta with <strong>nothing</strong></li>
</ul>
<p>&nbsp;</p>
</li>
<li>
<p>If the name of the GFF file does not match with any of the names of 
  the FASTA files, then the GFF file will be associated with all the FASTA
  files, considering that the FASTA files correspond to different 
  subgroups of the same dataset.</p>
<pre><code>orfold -fna sequences_Y.fasta sequences_X.fasta -options H -gff sequences_B.gff
</code></pre>
<p>ORFold associates:</p>
<ul>
<li>sequences_Y.fasta with sequences_B.gff</li>
<li>sequences_X.fasta with sequences_B.gff</li>
</ul>
<p>&nbsp;</p>
</li>
</ul>
</li>
<li>
<p>If the user provides multiple FASTA and GFF files (but not the same number)
   , all GFF files must have a corresponding FASTA file with the same root name. 
   Otherwise, ORFold will give an ERROR message. </p>
<pre><code>orfold -fna sequences_Y.fasta sequences_X.fasta sequences_Z.fasta -options H -gff sequences_Z.fasta sequences_Y.gff
</code></pre>
<p>ORFold associates:</p>
<ul>
<li>sequences_Y.fasta with sequences_Y.gff</li>
<li>sequences_Z.fasta with sequences_Z.gff</li>
<li>sequences_X.fasta with nothing</li>
</ul>
<p>&nbsp;</p>
<pre><code>orfold -fna sequences_Y.fasta sequences_X.fasta sequences_Z.fasta -options H -gff sequences_B.fasta sequences_A.gff
</code></pre>
<p>ORFold will give the following error message:</p>
<pre><code>Oups! You provided GFF file(s) which has/have no correspondance to the input FASTA files
</code></pre>
</li>
</ol>
<h3 id="running-orfold-on-subsets-of-randomly-selected-sequences">Running ORFold on subsets of randomly selected sequences<a class="headerlink" href="#running-orfold-on-subsets-of-randomly-selected-sequences" title="Permanent link">#</a></h3>
<p>Working with complete genomes could generate big amounts of sequences 
which can dramatically increase the computational time of ORFold when dealing with
large genomes (especially if the estimation of the disorder and aggregation propensities are
activated). If the user does not want to treat all 
the sequences, he can create a random sample of the input sequences, large enough
 to have an estimation of the distribution of the studied properties of its 
dataset from a representative sample of the input sequences. To do so, 
the user must indicate the number of sequences that are to be randomly selected
with the <strong>-N</strong> option. For a representative dataset, we recommend selecting at least
10000 sequences.</p>
<pre><code>orfold -fna sequences.fasta -options HIT -gff sequences.gff -N 10000
</code></pre>
<p>In this example, ORFold will estimate the fold potential, and the disorder and 
aggregation propensities on a sample of 10000 sequences extracted randomly 
from the initial <strong>sequences.fasta</strong> file.    </p>
<div class="admonition note">
    <p class="first admonition-title">
        Note
    </p>
    <p class="last">

    If the user works with more than one FASTA file and wishes to create 
    random samples for all the input sequence files, he has to indicate in 
    the -N option the size for each input file explicitly (in the same order 
    as the inputs passed in the -fna option).
    <br>

<pre><code>orfold -fna sequences_X.pfasta sequences_Y.pfasta -options H -N 1500 3000
</code></pre>

Also, if the user wants to sample two subsets of same sizes, he has to indicate the subset sizes explicitly for each input

<pre><code>orfold -fna sequences_X.pfasta sequences_Y.pfasta -options H -N 1500 1500
</code></pre>

    If the user whishes to calculate the fold potential of all the sequences 
    of one of the given inputs, he has to indicate it with the "all" flag (again with respect to
    the order of input files)

<pre><code class="language-bash">orfold -fna sequences_X.pfasta sequences_Y.pfasta -options H -N all 3000
</code></pre>

In this case, ORFold will calculate the fold potential for <b>all</b> the sequences 
in the sequences_X file while will generate a random sample of 3000 sequences for the 
sequences_Y file.   

    </p>
</div>

<p>References</p>
<ol>
<li>Robinson JT, Thorvaldsdóttir H, Winckler W, et al (2011)
   Integrative genomics viewer. Nature biotechnology 29:24–26</li>
</ol>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="parameters_orfold.html" class="btn btn-neutral float-right" title="ORFold parameters">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="Plot_orfold.html" class="btn btn-neutral" title="Plots with ORFplot"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="Plot_orfold.html" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="parameters_orfold.html" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '.';</script>
    <script src="js/theme.js" defer></script>
      <script src="search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
