## Usage
Here is presented an example of the use of ORFmap with ORFold 
to annotate all the ORFs of a genome, and to characterize their fold potential
and disorder and aggregation propensities.

The inputs used in this example and the resulting outputs are 
available in the ORFmine/examples/ folder.


## Annotation and extraction of ORFs with ORfmap

### Annotation of ORFs
ORFmap takes as input (i) a FASTA file containing the nucleotide
sequences of all chromosomes or contigs, and (ii) their corresponding 
annotations in a GFF file (see https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md
for more details on GFF3 format). 
The following command runs ORFmap on the complete genome of Escherichia
coli (Ecoli) available in the "examples" folder:

``` python
orfmap xxxx
```
ORFmap generates a novel GFF file (available in 
ORFmine/examples/outputs/Ecoli_ORFs_XXXX.gff)
containing the annotations of all identified ORFs. 

### Extraction of noncoding ORFs in a FASTA file
The amino acid sequences of 
all annotated ORFs or specific subsets of ORFs (i.e. only noncoding 
intergenic ORFs for example) can be extracted and written in a 
FASTA file with ORFget, a tool provided with ORFmap. The following 
instruction writes the amino acid sequences of all noncoding ORFs 
of Ecoli in a FASTA file.


``` python
orfget xxxx
```

ORFget produces a FASTA file containing the amino acid sequences
of all noncoding ORFs annotated by ORFmap (see xxxx for more examples 
on the use of ORFget). 

### Extraction of protein sequences of Ecoli in a FASTA file
In addition, ORFget enables the reconstruction of all protein 
sequences of Ecoli according to their definition in the original
GFF file.


``` python
orfget build CDS
```



## Characterization of the fold potential and disorder and aggregation propensities of the noncoding ORFs with ORFold


### Running ORFold
ORFold only needs a FASTA file containing the amino acid
sequences to characterize. That said, a GFF 
file containing the annotation of the input sequences can be provided
as well. In the latter case, ORFold writes the fold potential
and disorder and aggregation propensities in 3 GFF files respectively.
The latter can be subsequently uploaded on a genome viewer for a 
visual inspection of the distribution of these properties along
the genome's ORFs. 
The following instruction enables the user to probe the fold 
potential and disorder and aggregation propensities of all noncoding
ORFs annotated by ORFmap along with those of the protein sequences
of Ecoli. The GFF file generated by ORFmap is given as well in order
to manually inspect with a genome viewer, the distribution of
the fold potential and disorder and aggregation propensities
 of the noncoding ORFs along the genome of Ecoli.

``` python
orfold xxxx NONCODING + CDS 
```

ORFold provides 6 tables (3 per input dataset available in 
ORFmine/examples/outputs/). Each table contains for each input 
sequence, the fold potential, disorder and aggregation propensities
based on HCA [REF], Iupred [REF] and TANGO [REF] respectively 
(see here XXX for more details on the calculation of these scores).
ORFold also generates 3 novel GFF files containing the fold potential
and disorder and aggregation propensities respectively. The output
GFF files can be subsequently uploaded on a genome viewer.


### Visualization of the fold potential distribution with ORFplot

ORFplot enables the visualization of the distribution of the fold
potential of the noncoding ORFs and protein sequences along with the one of a reference dataset
of globular proteins taken from XXXX.

``` python
orfplot xxxx
```

The output graphic is generated in PNG/PDF format (see here XXX for
other examples of use of ORFplot).  


<div class="admonition note">
    <p class="first admonition-title">
        Note
    </p>
    <p class="last">
        This section implies that Iupred and TANGO have been downloaded 
and that their paths have been correctly defined during the installation 
procedure. 
    </p>
</div>
